### PR TITLE
Move from google test to catch2

### LINKED ISSUE #58: Drop Google test in favor of Catch2
`Google Test` from the next releases will start to depend on `abseil`, see [here](https://github.com/google/googletest/blob/master/README.md#coming-soon). `Abseil` is a quite big  C++ library that is not packaged in Ubuntu or Homebrew (see https://repology.org/project/abseil-cpp/versions), and will make `Google Test` a non-trivial dependency to obtain. On light of this I would propose moving from `Google Test` to `Catch2`.

@traversaro and @S-Dafarra

### PR DESCRIPTION
**Pull request by GiulioRomualdi**
This PR closes #58 

Continuous integration will fail. As soon as we merge the PR. I will work on #30

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -20,15 +23,6 @@ list(APPEND CMAKE_MODULE_PATH ${CMAKE_CURRENT_SOURCE_DIR}/cmake)
 
 option(BUILD_SHARED_LIBS "Build libraries as shared as opposed to static" ON)
 
-# Enable C++14
-set(CMAKE_CXX_EXTENSIONS OFF)
-set(CMAKE_CXX_STANDARD 14)
-set(CMAKE_CXX_STANDARD_REQUIRED 11)
```
Comment (S-Dafarra) [inline CMakeLists.txt:26]: Why did you remove this? Btw, it should be ``set(CMAKE_CXX_STANDARD_REQUIRED ON)``

File: cmake/AddOsqpEigenUnitTest.cmake [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+      set(unit_test_files ${${prefix}_SOURCES})
+      set(depends_on_eigen_private ${${prefix}_DEPENDS_ON_EIGEN_PRIVATE})
+
+      set(targetname ${name}UnitTests)
+      add_executable(${targetname}
+          "${unit_test_files}")
+
+      target_link_libraries(${targetname} PRIVATE CatchTestMain ${${prefix}_LINKS})
+      target_compile_definitions(${targetname} PRIVATE CATCH_CONFIG_FAST_COMPILE CATCH_CONFIG_DISABLE_MATCHERS)
+      target_compile_features(${targetname} PUBLIC cxx_std_17)
```
Comment (S-Dafarra) [inline cmake/AddOsqpEigenUnitTest.cmake:70]: Is this necessary?

File: cmake/OsqpEigenFindDependencies.cmake [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+
+endmacro()
+
+
+
+################################################################################
+# Find all packages
+
+find_package(Eigen3 3.2.92 REQUIRED)
+find_package(osqp REQUIRED)
```
Comment (S-Dafarra) [inline cmake/OsqpEigenFindDependencies.cmake:123]: Maybe we can move the definition of dependencies to a separate file to avoid mixing ``CMake`` logic and actual dependencies :thinking:

File: tests/CMakeLists.txt [hunk 4]
```diff
@@ -1,64 +1,27 @@
...
+
+add_osqpeigen_test(
+  NAME MPC
+  SOURCES MPCTest.cpp
+  LINKS OsqpEigen::OsqpEigen Eigen3::Eigen)
+
+add_osqpeigen_test(
+  NAME MPCUpdateMatrices
+  SOURCES MPCUpdateMatricesTest.cpp
+  LINKS OsqpEigen::OsqpEigen Eigen3::Eigen)
```
Comment (S-Dafarra) [inline tests/CMakeLists.txt:27]: Should the tests link ``Eigen3::Eigen``? Was there any issue in linking ``Eigen3::Eigen`` publicly in ``OsqpEigen``?

File: cmake/AddOsqpEigenUnitTest.cmake [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+    add_library(CatchTestMain ${CMAKE_BINARY_DIR}/Testing/Catch2Main.cpp)
+    target_link_libraries(CatchTestMain PUBLIC Catch2::Catch2)
+endif()
+
+
+function(add_osqpeigen_test)
+
+    if(OSQPEIGEN_COMPILE_tests)
+
+      set(options DEPENDS_ON_EIGEN_PRIVATE)
```
Comment (S-Dafarra) [inline cmake/AddOsqpEigenUnitTest.cmake:48]: Is this necessary?

File: cmake/AddOsqpEigenUnitTest.cmake (hunk 2 above)
Comment (GiulioRomualdi) [reply cmake/AddOsqpEigenUnitTest.cmake:70]: `cxx_std_17` is not necessary. I can change it with `cxx_std_14`

File: tests/CMakeLists.txt (hunk 4 above)
Comment (GiulioRomualdi) [reply tests/CMakeLists.txt:27]: I remember that back in time @traversaro suggested avoiding linking  `Eigen3::Eigen` publicly. Probably now something changed and we can link it

Comment (GiulioRomualdi) [reply tests/CMakeLists.txt:27]: Actually I think we can also remove `FindEigen3.cmake` file. Right?

Comment (traversaro) [reply tests/CMakeLists.txt:27]: Yes, the problem with linking and installed target as a public dependency with `Eigen3::Eigen`  was that then you need  add `find_depedency(Eigen3 CONFIG REQUIRED)`  in your `OSQPEigen-config.cmake`  file. Back in time, we had to support platforms (i.e. the legacy Windows binaries) that contained Eigen but without distributing a `Eigen3Config.cmake`, so that line was not working out of the box. I think the problem is not there anymore, so we can just link as a public dependency  without problems.

Comment (traversaro) [reply tests/CMakeLists.txt:27]: > 
> Actually I think we can also remove `FindEigen3.cmake` file. Right?

Yes!

File: cmake/AddOsqpEigenUnitTest.cmake (hunk 5 above)
Comment (GiulioRomualdi) [reply cmake/AddOsqpEigenUnitTest.cmake:48]: I don't think so

File: tests/CMakeLists.txt (hunk 4 above)
Comment (S-Dafarra) [reply tests/CMakeLists.txt:27]: > Yes, the problem with linking and installed target as a public dependency with Eigen3::Eigen was that then you need add find_depedency(Eigen3 CONFIG REQUIRED) in your OSQPEigen-config.cmake file. Back in time, we had to support platforms (i.e. the legacy Windows binaries) that contained Eigen but without distributing a Eigen3Config.cmake, so that line was not working out of the box. I think the problem is not there anymore, so we can just link as a public dependency without problems.

Thanks! I did not remember that. I thought the problem was mainly in the inclusion in public headers.

File: CMakeLists.txt (hunk 1 above)
Comment (GiulioRomualdi) [reply CMakeLists.txt:26]: `set(CMAKE_CXX_EXTENSIONS OFF)` should not be removed (it's my fault).

While
```cmake
set(CMAKE_CXX_STANDARD 14)
set(CMAKE_CXX_STANDARD_REQUIRED 11)
```
can be removed if I add `target_compile_features(${LIBRARY_TARGET_NAME} PUBLIC cxx_std_14)`. If I'm not mistaken adding this kind of commands may lead to problem when the library is compiled within other projects (@traversaro  may confirm/deny)

File: cmake/AddOsqpEigenUnitTest.cmake (hunk 5 above)
Comment (GiulioRomualdi) [reply cmake/AddOsqpEigenUnitTest.cmake]: Removed in 24b77cac3a19e43ae2bbf663b2ee4a450d5130ca

File: cmake/OsqpEigenFindDependencies.cmake (hunk 3 above)
Comment (GiulioRomualdi) [reply cmake/OsqpEigenFindDependencies.cmake]: Done in db75a1857f7f3aba9fe3227bdc2465be3fdb37cf

File: cmake/AddOsqpEigenUnitTest.cmake (hunk 2 above)
Comment (GiulioRomualdi) [reply cmake/AddOsqpEigenUnitTest.cmake]: Done in 24b77cac3a19e43ae2bbf663b2ee4a450d5130ca

File: CMakeLists.txt (hunk 1 above)
Comment (GiulioRomualdi) [reply CMakeLists.txt]: In 24b77cac3a19e43ae2bbf663b2ee4a450d5130ca i restored `set(CMAKE_CXX_EXTENSIONS OFF)` and  in 69a410aeab45b280500b5f13941f38bab6d48c05 i required `c++14`  for compiling the library

**Comment by GiulioRomualdi** [comment]
In https://github.com/robotology/osqp-eigen/pull/59/commits/5c27d0e42c2ee413d64dfd548c808bd0ba38fa32 I implemented the possibility to add customs `target_compile_definitions` for the UnitTests. This is required to make the tests compiling on windows (Suggested by @traversaro  in https://github.com/robotology/osqp-eigen/pull/30#issuecomment-628571179)

File: tests/CMakeLists.txt (hunk 4 above)
Comment (S-Dafarra) [reply tests/CMakeLists.txt]: If we link ``Eigen`` as a public dependency, I guess it should be added also in https://github.com/robotology/osqp-eigen/blob/50154d5047a38b5b68b3a8994a6a1e76effabf36/CMakeLists.txt#L97
right?

Comment (GiulioRomualdi) [reply tests/CMakeLists.txt]: I think so

Comment (GiulioRomualdi) [reply tests/CMakeLists.txt]: In bc4b7845aa05a77e525d619a9e5d896cd3e4a62b `Eigen3` is now linked as `public` library
